File: frontend/src/lib/domain/nutrition.ts
/**
 * Nutrition calculation engine — pure, deterministic, side-effect free.
 *
 * All business rules for turning (food, quantity, unit) into nutrients and
 * aggregating them live HERE, never in components. Phase 2 mirrors this
 * module in Python; both run the same test vectors (see nutrition.test.ts).
 *
 * Rounding policy (explicit, applied nowhere else):
 * - Internal math is full-precision floating point.
 * - `roundNutrients` is applied only at persistence/presentation boundaries:
 *   kcal → nearest integer; macros/micros → 1 decimal; mg values → nearest integer.
 * - Half-up rounding (0.5 → 1), matching what users expect from labels.
 */

import type { Food, Nutrients, RecipeIngredient } from '../api/types';

// ---------- Rounding ----------

/** Deterministic half-up rounding to `dp` decimals (avoids JS toFixed quirks). */
export function roundHalfUp(value: number, dp = 0): number {
  const f = 10 ** dp;
  // EPSILON nudge so values like 2.675 (stored as 2.67499...) round as printed
  return Math.round((value + Number.EPSILON * Math.sign(value)) * f) / f;
}

const INT_FIELDS: (keyof Nutrients)[] = ['kcal', 'sodiumMg', 'potassiumMg', 'calciumMg', 'cholesterolMg'];

export function roundNutrients(n: Nutrients): Nutrients {
  const out = {} as Record<string, number>;
  for (const [k, v] of Object.entries(n)) {
    if (v === undefined || v === null) continue;
    out[k] = roundHalfUp(v, INT_FIELDS.includes(k as keyof Nutrients) ? 0 : 1);
  }
  return out as unknown as Nutrients;
}

// ---------- Unit resolution ----------

/**
 * Resolve a logged (quantity, unit) to grams.
 * `unitId === 'g'` means raw grams (or ml for liquids — same 1:1 basis).
 * Throws on unknown unit ids or non-finite/negative quantities: the caller
 * (form validation) should have prevented both, and silently guessing would
 * corrupt health records.
 */
export function resolveGrams(food: Pick<Food, 'servingUnits'>, quantity: number, unitId: string): number {
  if (!Number.isFinite(quantity) || quantity < 0) {
    throw new RangeError(`Invalid quantity: ${quantity}`);
  }
  if (unitId === 'g') return quantity;
  const unit = food.servingUnits.find((u) => u.id === unitId);
  if (!unit) throw new RangeError(`Unknown serving unit '${unitId}'`);
  return quantity * unit.grams;
}

// ---------- Scaling ----------

/**
 * Scale per-100g nutrients to a gram weight. Missing optional nutrients stay
 * missing (undefined) — we never invent zeros for unknown micronutrient data,
 * so totals can distinguish "0 mg" from "not measured".
 */
export function scaleNutrients(per100: Nutrients, grams: number): Nutrients {
  const factor = grams / 100;
  const out = {} as Record<string, number>;
  for (const [k, v] of Object.entries(per100)) {
    if (v === undefined || v === null) continue;
    out[k] = v * factor;
  }
  return out as unknown as Nutrients;
}

/** Nutrients for logging `quantity` × `unitId` of `food`. Rounded for persistence. */
export function nutrientsForServing(food: Food, quantity: number, unitId: string): { grams: number; nutrients: Nutrients } {
  const grams = resolveGrams(food, quantity, unitId);
  return { grams, nutrients: roundNutrients(scaleNutrients(food.nutrients, grams)) };
}

// ---------- Aggregation ----------

const EMPTY: Nutrients = { kcal: 0, proteinG: 0, carbsG: 0, fatG: 0 };

/**
 * Sum nutrient snapshots. Core macros always sum (default 0);
 * optional micros sum only when at least one input reports them.
 */
export function sumNutrients(items: Nutrients[]): Nutrients {
  const out: Record<string, number> = { ...EMPTY };
  for (const n of items) {
    for (const [k, v] of Object.entries(n)) {
      if (v === undefined || v === null) continue;
      out[k] = (out[k] ?? 0) + v;
    }
  }
  return out as unknown as Nutrients;
}

// ---------- Recipes ----------

/**
 * Per-serving nutrients of a recipe: sum of resolved ingredient snapshots ÷ servings.
 * Ingredient lines carry their own resolved `grams` and reference an immutable
 * food version, so this is reproducible even after the source food is corrected.
 */
export function recipePerServing(
  ingredients: Array<Pick<RecipeIngredient, 'grams'> & { per100: Nutrients }>,
  servings: number,
): Nutrients {
  if (!Number.isFinite(servings) || servings <= 0) {
    throw new RangeError(`Recipe servings must be > 0, got ${servings}`);
  }
  const total = sumNutrients(ingredients.map((i) => scaleNutrients(i.per100, i.grams)));
  const perServing: Record<string, number> = {};
  for (const [k, v] of Object.entries(total)) perServing[k] = (v as number) / servings;
  return roundNutrients(perServing as unknown as Nutrients);
}

// ---------- Daily budget ----------

export interface DayBudgetInput {
  calorieTarget: number;
  consumedKcal: number;
  exerciseBurnedKcal: number;
  /** whether exercise burn credits back into the budget (user setting) */
  creditExercise: boolean;
}

export function remainingKcal({ calorieTarget, consumedKcal, exerciseBurnedKcal, creditExercise }: DayBudgetInput): number {
  const credit = creditExercise ? exerciseBurnedKcal : 0;
  return roundHalfUp(calorieTarget - consumedKcal + credit);
}

/** Progress toward a target, clamped to [0, 1] for rings/bars. Never NaN. */
export function progressPct(value: number, target: number): number {
  if (!Number.isFinite(target) || target <= 0) return 0;
  return Math.min(1, Math.max(0, value / target));
}

// ---------- Display helpers (pure formatting, no locale side effects) ----------

export function formatGrams(g: number): string {
  return `${roundHalfUp(g, g < 10 ? 1 : 0)} g`;
}

export function formatKcal(kcal: number): string {
  return `${roundHalfUp(kcal).toLocaleString('en-US')} kcal`;
}

File: backend/app/domain/nutrition.py
"""Nutrition engine — Python mirror of frontend/src/lib/domain/nutrition.ts.

Both implementations run the same test vectors (tests/test_nutrition.py mirrors
nutrition.test.ts). Rounding policy: half-up; kcal & mg fields to integers,
gram fields to 1 decimal; applied only at persistence/presentation boundaries.
"""

from __future__ import annotations

import math
from typing import Any

Nutrients = dict[str, float]

INT_FIELDS = {"kcal", "sodiumMg", "potassiumMg", "calciumMg", "cholesterolMg"}


def round_half_up(value: float, dp: int = 0) -> float:
    f = 10**dp
    nudge = math.copysign(2.220446049250313e-16, value)  # EPSILON, mirrors JS
    out = math.floor(abs((value + nudge) * f) + 0.5) * math.copysign(1, value) / f
    return int(out) if dp == 0 else out


def round_nutrients(n: Nutrients) -> Nutrients:
    return {
        k: round_half_up(v, 0 if k in INT_FIELDS else 1)
        for k, v in n.items()
        if v is not None
    }


def resolve_grams(serving_units: list[dict[str, Any]], quantity: float, unit_id: str) -> float:
    if not math.isfinite(quantity) or quantity < 0:
        raise ValueError(f"Invalid quantity: {quantity}")
    if unit_id == "g":
        return quantity
    for u in serving_units:
        if u["id"] == unit_id:
            return quantity * u["grams"]
    raise ValueError(f"Unknown serving unit '{unit_id}'")


def scale_nutrients(per100: Nutrients, grams: float) -> Nutrients:
    factor = grams / 100
    return {k: v * factor for k, v in per100.items() if v is not None}


def nutrients_for_serving(
    per100: Nutrients, serving_units: list[dict[str, Any]], quantity: float, unit_id: str
) -> tuple[float, Nutrients]:
    grams = resolve_grams(serving_units, quantity, unit_id)
    return grams, round_nutrients(scale_nutrients(per100, grams))


EMPTY: Nutrients = {"kcal": 0, "proteinG": 0, "carbsG": 0, "fatG": 0}


def sum_nutrients(items: list[Nutrients]) -> Nutrients:
    out: Nutrients = dict(EMPTY)
    for n in items:
        for k, v in n.items():
            if v is None:
                continue
            out[k] = out.get(k, 0) + v
    return out


def recipe_per_serving(ingredients: list[dict[str, Any]], servings: float) -> Nutrients:
    """ingredients: [{grams, per100}]"""
    if not math.isfinite(servings) or servings <= 0:
        raise ValueError(f"Recipe servings must be > 0, got {servings}")
    total = sum_nutrients([scale_nutrients(i["per100"], i["grams"]) for i in ingredients])
    return round_nutrients({k: v / servings for k, v in total.items()})


def remaining_kcal(calorie_target: float, consumed_kcal: float, exercise_burned: float, credit_exercise: bool = True) -> float:
    credit = exercise_burned if credit_exercise else 0
    return round_half_up(calorie_target - consumed_kcal + credit)
